crypto/fipsmodule/ec/p521.c

```diff
@@ -0,0 +1,1002 @@
+
+
+// dkostic: these macro definitions will make sense once we add s2n-bignum
+// functions as in p384.c
+// Fiat-crypto implementation of field arithmetic
+#define p521_felem_add(out, in0, in1)   fiat_secp521r1_carry_add(out, in0, in1)
+#define p521_felem_sub(out, in0, in1)   fiat_secp521r1_carry_sub(out, in0, in1)
+#define p521_felem_opp(out, in0)        fiat_secp521r1_carry_opp(out, in0)
+#define p521_felem_mul(out, in0, in1)   fiat_secp521r1_carry_mul(out, in0, in1)
+#define p521_felem_sqr(out, in0)        fiat_secp521r1_carry_square(out, in0)
+#define p521_felem_to_bytes(out, in0)   fiat_secp521r1_to_bytes(out, in0)
+#define p521_felem_from_bytes(out, in0) fiat_secp521r1_from_bytes(out, in0)
+
+static p521_limb_t p521_felem_nz(const p521_limb_t in1[P521_NLIMBS]) {
+  p521_limb_t is_not_zero = 0;
+  for (int i = 0; i < P521_NLIMBS; i++) {
+    is_not_zero |= in1[i];
+  }
+  // Fiat-crypto functions may return p (the field characteristic)
+  // instead of 0 in some cases, so we also check for that.
+  p521_limb_t is_not_p = 0;
+  for (int i = 0; i < P521_NLIMBS; i++) {
+    is_not_p |= (in1[i] ^ p521_felem_p[i]);
+  }
+
+  return (is_not_zero && is_not_p);
+}
+
+static void p521_felem_copy(p521_limb_t out[P521_NLIMBS],
+                            const p521_limb_t in1[P521_NLIMBS]) {
+  for (size_t i = 0; i < P521_NLIMBS; i++) {
+    out[i] = in1[i];
+  }
+}
+
+static void p521_felem_cmovznz(p521_limb_t out[P521_NLIMBS],
+                               p521_limb_t t,
+                               const p521_limb_t z[P521_NLIMBS],
+                               const p521_limb_t nz[P521_NLIMBS]) {
+  fiat_secp521r1_selectznz(out, !!t, z, nz);
+}
+
+static void p521_from_generic(p521_felem out, const EC_FELEM *in) {
+  p521_felem_from_bytes(out, in->bytes);
+}
+
+static void p521_to_generic(EC_FELEM *out, const p521_felem in) {
+  // |p521_felem_to_bytes| function will write the result to the first 66 bytes
+  // of |out| which is exactly how many bytes are needed to represent a 521-bit
+  // element. However, EC_FELEM is a union of uint8_t array and BN_ULONG array.
+  // The number of BN_ULONGs to represent a 521-bit value is 9 and 17, when
+  // BN_ULONG is 64-bit and 32-bit, respectively. Nine 64-bit BN_ULONGs
+  // translate to 72 bytes, which means that we have to make sure that the
+  // extra 6 bytes are zeroed out. To avoid confusion over 32 vs. 64 bit
+  // systems and Fiat's vs. ours representation we zero out the whole element.
+  OPENSSL_memset((uint8_t*)out->words, 0, sizeof(out->words));
+  // Convert the element to bytes.
+  p521_felem_to_bytes(out->bytes, in);
+}
+
+// Finite field inversion using Fermat Little Theorem.
+// The code is autogenerated by the ECCKiila project:
+//   https://arxiv.org/abs/2007.11481
+static void p521_felem_inv(p521_felem output, const p521_felem t1) {
+    /* temporary variables */
+    p521_felem acc, t128, t16, t2, t256, t32, t4;
+    p521_felem t512, t516, t518, t519, t64, t8;
```

Comment:
Nit: we can rearrange the variable names in their order of appearance, which I think is their ascending order, unless you'd like to keep it true to where it's copied from.

Reply:
done